godot-ffi/src/godot_ffi.rs

```diff
@@ -487,3 +487,67 @@ mod scalars {
         }
     }
 }
+
+unsafe impl<T> GodotFfi for Option<T>
+where
+    T: GodotFfi,
+{
+    fn sys(&self) -> sys::GDExtensionTypePtr {
+        match self {
+            Some(value) => value.sys(),
+            None => {
+                let nil_ptr = ptr::null_mut();
+                let new_nil = interface_fn!(variant_new_nil);
+
+                unsafe {
+                    new_nil(nil_ptr);
+                }
+
+                nil_ptr as sys::GDExtensionTypePtr
+            }
+        }
+    }
+
+    unsafe fn from_sys(ptr: sys::GDExtensionTypePtr) -> Self {
+        let nil_ptr = ptr::null_mut();
+
+        interface_fn!(variant_new_nil)(nil_ptr);
+
+        if ptr as sys::GDExtensionVariantPtr == nil_ptr {
+            return None;
+        }
+
+        Some(T::from_sys(ptr))
+    }
+
+    unsafe fn from_sys_init(init_fn: impl FnOnce(sys::GDExtensionTypePtr)) -> Self {
+        let mut raw = std::mem::MaybeUninit::uninit();
+        init_fn(raw.as_mut_ptr() as sys::GDExtensionTypePtr);
+
+        Self::from_sys(raw.assume_init())
+    }
+
+    unsafe fn from_arg_ptr(ptr: sys::GDExtensionTypePtr, call_type: PtrcallType) -> Self {
+        let nil_ptr = ptr::null_mut();
+
+        interface_fn!(variant_new_nil)(nil_ptr);
+
+        if ptr as sys::GDExtensionVariantPtr == nil_ptr {
+            return None;
+        }
+
+        Some(T::from_arg_ptr(ptr, call_type))
+    }
+
+    unsafe fn move_return_ptr(self, ptr: sys::GDExtensionTypePtr, call_type: PtrcallType) {
+        match self {
+            Some(gd) => gd.move_return_ptr(ptr, call_type),
+            None => {
+                let nil_ptr = ptr::null_mut();
+
+                interface_fn!(variant_new_nil)(nil_ptr);
+                std::ptr::write(ptr as *mut _, nil_ptr as sys::GDExtensionTypePtr)
```

Comment:
im not sure if this is correct for all types. as you're here writing a null pointer into whatever is being pointed at by `ptr`, but that value may be bigger than a pointer.

Reply:
You are completely right. After doing some manual tests, it became clear to me that it was nonsense. I revised the implementation, and now it does work for the types I verified so far (`Node3D`, `i64`, `GodotString`). Primitive types are not nullable in Godot / GDScript, though. At the moment, Godot coverts them to their default value when passing the nil variant to GDScript, i.e. empty-string or `0`, so I think `Option<T>` should only be available as `Option<Gd<T>>`